### PR TITLE
Implementation of Tiled Images + tests

### PR DESCRIPTION
**Pull request by msft-Jeyaram**
This change is [Reviewable](https://reviewable.io/reviews/microsoft/winobjc/1542)

### CODE REVIEW

**Comment by msft-Jeyaram** [comment]
Commit 	769d982

File: Frameworks/CoreGraphics/CGContext.mm [hunk 1]
```diff
@@ -1950,30 +1950,43 @@ void CGContextEOFillPath(CGContextRef context) {
...
 
-    // Obtain the pixel format for the image
-    WICPixelFormatGUID imagePixelFormat = _CGImageGetWICPixelFormat(image);
+    ComPtr<IWICBitmap> bmap;
+    RETURN_IF_FAILED(_CGImageGetWICImageSource(image, &bmap));
 
-    CFRetain(image);
-    woc::unique_cf<CGImageRef> refImage;
-    refImage.reset(image);
+    ComPtr<ID2D1Bitmap> d2dBitmap;
```
Comment (DHowett-MSFT) [inline Frameworks/CoreGraphics/CGContext.mm]: This requires a conversion into a valid render target format first.

There is some work around this in #1528.

Comment (DHowett-MSFT) [inline Frameworks/CoreGraphics/CGContext.mm]: Oh, I see. I don't think this needs to be in two functions :P They'll always be found together.

File: Frameworks/CoreGraphics/CGContext.mm [hunk 2]
```diff
@@ -1950,30 +1950,43 @@ void CGContextEOFillPath(CGContextRef context) {
...
+    WICPixelFormatGUID imagePixelFormat = _CGImageGetWICPixelFormat(image);
     if (!_CGIsValidRenderTargetPixelFormat(imagePixelFormat)) {
         // convert it to a valid pixelformat
-        refImage.reset(_CGImageCreateCopyWithPixelFormat(image, GUID_WICPixelFormat32bppPBGRA));
+        return _CGImageCreateCopyWithPixelFormat(image, GUID_WICPixelFormat32bppPBGRA);
     }
 
-    ComPtr<IWICBitmap> bmap;
-    FAIL_FAST_IF_FAILED(_CGImageGetWICImageSource(refImage.get(), &bmap));
+    CGImageRetain(image);
```
Comment (aballway) [inline Frameworks/CoreGraphics/CGContext.mm]: nit: these can be one line

File: Frameworks/CoreGraphics/CGContext.mm [hunk 3]
```diff
@@ -1950,30 +1950,43 @@ void CGContextEOFillPath(CGContextRef context) {
...
+    return image;
+}
+
+/**
+ @Status Interoperable
+*/
+void CGContextDrawImage(CGContextRef context, CGRect rect, CGImageRef image) {
+    NOISY_RETURN_IF_NULL(context);
+    NOISY_RETURN_IF_NULL(image);
```
Comment (aballway) [inline Frameworks/CoreGraphics/CGContext.mm]: nit: should this be noisy or a simple no-op?

File: Frameworks/CoreGraphics/CGContext.mm [hunk 4]
```diff
@@ -2022,13 +2035,58 @@ void _CGContextDrawImageRect(CGContextRef context, CGImageRef image, CGRect src,
...
+    // |0 -1 0|
+    // |0 2m 1|
+    //
+    // Combined with [scale sx * sy] * [translate X, Y], that becomes:
+    // |sx     0 0|
+    // | 0   -sy 0|
+    // | x -y+2m 0|
+    // Or, the transformation matrix for drawing a flipped rect at a scale and offset.
+    D2D1_SIZE_U bitmapSize = d2dBitmap->GetPixelSize();
+    CGFloat sx = rect.size.width / bitmapSize.width;
```
Comment (aballway) [inline Frameworks/CoreGraphics/CGContext.mm]: check that width, height != 0

File: Frameworks/CoreGraphics/CGContext.mm (hunk 3 above)
Comment (msft-Jeyaram) [inline Frameworks/CoreGraphics/CGContext.mm]: noisy

---
In reply to: [92660158](https://github.com/Microsoft/WinObjC/pull/1542#discussion_r92660158) [](ancestors = 92660158)

File: Frameworks/CoreGraphics/CGContext.mm [hunk 5]
```diff
@@ -2022,13 +2035,65 @@ void _CGContextDrawImageRect(CGContextRef context, CGImageRef image, CGRect src,
...
+    D2D1_SIZE_U bitmapSize = d2dBitmap->GetPixelSize();
+    FAIL_FAST_IF(bitmapSize.width == 0);
+    FAIL_FAST_IF(bitmapSize.height == 0);
+
+    CGFloat sx = rect.size.width / bitmapSize.width;
+    CGFloat sy = rect.size.height / bitmapSize.height;
+    CGFloat m = rect.origin.y + (rect.size.height / 2.f);
+
+    CGAffineTransform transform{ sx, 0, 0, -sy, rect.origin.x, (2 * m) - rect.origin.y };
+    transform = CGAffineTransformConcat(transform, CGContextGetUserSpaceToDeviceSpaceTransform(context));
```
Comment (rajsesh-msft) [inline Frameworks/CoreGraphics/CGContext.mm]: I think I saw the same block of code (with identical comments) for transforms in @DHowett-MSFT's PR.  can we make a utility function out of this?

File: Frameworks/CoreGraphics/CGContext.mm [hunk 6]
```diff
@@ -2022,13 +2035,65 @@ void _CGContextDrawImageRect(CGContextRef context, CGImageRef image, CGRect src,
...
+
+    ComPtr<ID2D1BitmapBrush1> bitmapBrush;
+    ComPtr<ID2D1DeviceContext> deviceContext = context->DeviceContext();
+    FAIL_FAST_IF_FAILED(
+        deviceContext->CreateBitmapBrush(d2dBitmap.Get(),
+                                         D2D1::BitmapBrushProperties1(D2D1_EXTEND_MODE_WRAP,
+                                                                      D2D1_EXTEND_MODE_WRAP,
+                                                                      (CGImageGetShouldInterpolate(refImage.get()) ?
+                                                                           context->CurrentGState().bitmapInterpolationMode :
+                                                                           D2D1_INTERPOLATION_MODE_NEAREST_NEIGHBOR)),
```
Comment (rajsesh-msft) [inline Frameworks/CoreGraphics/CGContext.mm]: The interpolation mode was added as a member utility in another PR.

File: Frameworks/CoreGraphics/CGContext.mm (hunk 5 above)
Comment (msft-Jeyaram) [inline Frameworks/CoreGraphics/CGContext.mm]: Definitely., it's being used by all brushes that have image as source.

File: Frameworks/CoreGraphics/CGContext.mm (hunk 6 above)
Comment (msft-Jeyaram) [inline Frameworks/CoreGraphics/CGContext.mm]: I'll pull that commit in

**Comment by msft-Jeyaram** [comment]
tests failed due to green line tests not being disabled :(

**Pull request merged**